pkg/selector/matcher.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021 VMware
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package selector
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/vmware-tanzu/cartographer/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/cartographer/pkg/eval"
+)
+
+type Selectable interface {
+	GetLabels() map[string]string
+}
+
+type MatchError interface {
+	error
+	SelectorIndex() int
+}
+
+type selectorMatchError struct {
+	Err                  error
+	SelectingObjectIndex int
+}
+
+func (e selectorMatchError) Error() string {
+	return e.Err.Error()
+}
+
+func (e selectorMatchError) SelectorIndex() int {
+	return e.SelectingObjectIndex
+}
+
+// BestSelectorMatchIndices returns a slice of indices into the passed in `selectors` which match the
+// `selectable` with the most specificity. Any error processing a selector includes the index of the
+// offending Selector.
+func BestSelectorMatchIndices(selectable Selectable, selectors []v1alpha1.Selector) ([]int, MatchError) {
+	var matchingSelectorIndicesByScore = map[int][]int{}
```

Comment:
It seems like we're building out a whole map unnecessarily, no? Don't we really just want an array of the indices with the highest matching score? If we reach a new max score on line 88, we could trash the array and rebuild it.

I'm not too worried about the storage complexity, but I think it's a little less clear what's going on the way it's written right now.